packages/ember-metal/lib/meta.js

```diff
@@ -0,0 +1,287 @@
+// Remove "use strict"; from transpiled module until
+// https://bugs.webkit.org/show_bug.cgi?id=138038 is fixed
+//
+'REMOVE_USE_STRICT: true';
+
+import isEnabled from 'ember-metal/features';
+import { protoMethods as listenerMethods } from 'ember-metal/meta_listeners';
+
+/**
+@module ember-metal
+*/
+
+/*
+ This declares several meta-programmed members on the Meta class. Such
+ meta!
+
+ In general, the `readable` variants will give you an object (if it
+ already exists) that you can read but should not modify. The
+ `writable` variants will give you a mutable object, and they will
+ create it if it didn't already exist.
+
+ The following methods will get generated metaprogrammatically, and
+ I'm including them here for greppability:
+
+ writableCache, readableCache, writableWatching, readableWatching,
+ peekWatching, clearWatching, writableMixins, readableMixins,
+ peekMixins, clearMixins, writableBindings, readableBindings,
+ peekBindings, clearBindings, writableValues, readableValues,
+ peekValues, clearValues, writableDeps, readableDeps, getAllDeps
+ writableChainWatchers, readableChainWatchers, writableChains,
+ readableChains
+
+*/
+let members = {
+  cache: ownMap,
+  watching: inheritedMap,
+  mixins: inheritedMap,
+  bindings: inheritedMap,
+  values: inheritedMap,
+  deps: inheritedMapOfMaps,
+  chainWatchers: ownCustomObject,
+  chains: inheritedCustomObject
+};
+
+let memberNames = Object.keys(members);
+
+function Meta(obj, parentMeta) {
+  // preallocate a slot for each member
+  for (let i = 0; i < memberNames.length; i++) {
+    this[memberProperty(memberNames[i])] = undefined;
+  }
+
+  // used only internally
+  this.source = obj;
+
+  // when meta(obj).proto === obj, the object is intended to be only a
+  // prototype and doesn't need to actually be observable itself
+  this.proto = undefined;
+
+  // The next meta in our inheritance chain. We (will) track this
+  // explicitly instead of using prototypical inheritance because we
+  // have detailed knowledge of how each property should really be
+  // inherited, and we can optimize it much better than JS runtimes.
+  this.parent = parentMeta;
+
+  this._initializeListeners();
+}
+
+for (let name in listenerMethods) {
+  Meta.prototype[name] = listenerMethods[name];
+}
+
+(function setupMembers() {
+  for (let i = 0; i < memberNames.length; i++) {
+    let name = memberNames[i];
+    let implementation = members[name];
+    implementation(name, Meta);
+  }
+})();
+
+
+// Implements a member that is a lazily created, non-inheritable
+// POJO.
+function ownMap(name, Meta) {
+  let key = memberProperty(name);
+  let capitalized = capitalize(name);
+  Meta.prototype['writable' + capitalized] = function() {
+    return getOrCreateOwnMap.call(this, key);
+  };
+  Meta.prototype['readable' + capitalized] = function() { return this[key]; };
+}
+
+function getOrCreateOwnMap(key) {
+  let ret = this[key];
+  if (!ret) {
+    ret = this[key] = {};
```

Comment:
Yes, I actually started with `Object.create(null)` in all these places, but backed that out when I remembered how weirdly slow it is sometimes.

The above pattern avoids the performance hit of Object.create? If so, that seems bananas.